### DISCUSSION TITLE
Using Dask in FastAPI

### DISCUSSION THREAD

**Discussion by marcus302** [dask/dask#11272] [Q&A]
Hello there,

I'm trying to use Dask to process background jobs for my API, written in FastAPI. My current approach is:

1. Instantiating a local cluster on app startup.

```python
from contextlib import asynccontextmanager
from dask.distributed import LocalCluster

@asynccontextmanager
async def lifespan(app: FastAPI):
    # Load the dask cluster
    app.state.dask_cluster = LocalCluster(processes=False, asynchronous=True)
    yield
    # Clean up the ML models and release the resources
    close_call = app.state.dask_cluster.close()
    if close_call is not None:
        await close_call


app = FastAPI(lifespan=lifespan)
```

2. Creating a client and using it to submit jobs each time a route is called. I wait for the jobs to finish and return a response to the client.

```python
from fastapi import Request
from dask.distributed import Client, as_completed

@router.post(
    "/job",
    response_model=schemas.JobStartedConfirmation,
)
async def start_job(request: Request):
    async with Client(self.request.app.state.dask_cluster) as client:
        futures = []
        future_job = client.compute(start_new_job())
        futures.append(future_job)

        # Do more stuff, wait the futures, and return a response to the client.
```

Obviously waiting for the jobs to complete before returning means it's not a background job yet. I want to make sure the API responds immediately and Dask stays running in the background. I am wondering what the best approach / pattern is for this:

1. Should I also start a Client on startup and share it between different calls on the route?
2. Has what I'm trying to do already been done before? Where can I find documentation for this?

If anyone is willing to share their own solutions for something similar, I would be very grateful. I can document this and contribute to the project if it's welcome.

**Comment by swertz** [comment]
Here is one possible solution (I don't claim it is the best or the only one, this really depends on your requirements).

Create a `Variable` to hold the future after it has gone out of scope, and simply return a unique ID for the task when your `/job` endpoint is called. Then, provide another endpoint that takes that ID as input, and returns the status of the task or its result when it is done:

```python

@lru_cache
def dask_client():
    return Client(SCHEDULER_URI)

@app.post("/job")
async def start_job(dask_client: Annotated[dask.Client, Depends(dask_client)]):
    future_job = client.compute(start_new_job())
    
    task_uuid = str(uuid4().hex)
    var = Variable(task_uuid, dask_client)
    var.set(future)
    
    return {"task_uuid": task_uuid}

@app.post("/job/{task_uuid}")
async def get_job(task_uuid: str, dask_client: Annotated[dask.Client, Depends(dask_client)]):
    var = Variable(task_uuid, dask_client)
    future = var.get()

    if future.done():
        var.delete()
        return future.result()

    return {"task_status": future.state}
```

Of course you'll need more error handling and validation, but that gives you the overall idea. How does it sound?
